src/charm.py

```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charmed operator for the SDCORE AMF service."""
+
+import logging
+from ipaddress import IPv4Address
+from subprocess import check_output
+from typing import Union
+
+import ops
+from charms.data_platform_libs.v0.data_interfaces import DatabaseCreatedEvent, DatabaseRequires
+from charms.observability_libs.v1.kubernetes_service_patch import KubernetesServicePatch
+from charms.prometheus_k8s.v0.prometheus_scrape import (  # type: ignore[import]  # noqa: E501
+    MetricsEndpointProvider,
+)
+from charms.sdcore_nrf.v0.fiveg_nrf import NRFAvailableEvent, NRFRequires
+from jinja2 import Environment, FileSystemLoader
+from lightkube.models.core_v1 import ServicePort
+from ops.charm import CharmBase, PebbleReadyEvent
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+from ops.pebble import Layer
+
+logger = logging.getLogger(__name__)
+
+PROMETHEUS_PORT = 9089
+SBI_PORT = 29518
+NGAPP_PORT = 38412
+SCTP_GRPC_PORT = 9000
+AMF_DATABASE_NAME = "sdcore_amf"
+DEFAULT_DATABASE_NAME = "free5gc"
+CONFIG_DIR_PATH = "/free5gc/config"
+CONFIG_FILE_NAME = "amfcfg.conf"
+CONFIG_TEMPLATE_DIR_PATH = "src/templates/"
+CONFIG_TEMPLATE_NAME = "amfcfg.conf.j2"
+
+
+class AMFOperatorCharm(CharmBase):
+    """Main class to describe juju event handling for the 5G AMF operator."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._amf_container_name = self._amf_service_name = "amf"
+        self._amf_container = self.unit.get_container(self._amf_container_name)
+        self._nrf_requires = NRFRequires(charm=self, relation_name="nrf")
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._on_amf_pebble_ready)
+        self._amf_metrics_endpoint = MetricsEndpointProvider(self)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+                ServicePort(name="sbi", port=SBI_PORT),
+                ServicePort(name="ngapp", port=NGAPP_PORT, protocol="SCTP"),
+                ServicePort(name="sctp-grpc", port=SCTP_GRPC_PORT),
+            ],
+        )
+        self._amf_database = DatabaseRequires(
+            self, relation_name="amf-database", database_name=AMF_DATABASE_NAME
+        )
+        self._default_database = DatabaseRequires(
+            self, relation_name="default-database", database_name=DEFAULT_DATABASE_NAME
+        )
+        self.framework.observe(
+            self.on.default_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(self.on.amf_pebble_ready, self._on_amf_pebble_ready)
+
+    def _on_amf_pebble_ready(
+        self,
+        event: Union[PebbleReadyEvent, DatabaseCreatedEvent, NRFAvailableEvent],
```

Comment:
```suggestion
        event: EventBase,
```
We are not using anything on the event except `defer()`, so for typing it makes more sense to just expect any events.

Reply:
✅